### PR TITLE
fix #7997

### LINKED ISSUE #7997: Native tests in crossbuild
**Describe the bug**
In some cases useful is build and run tests on build system without any wrappers. It not possible by now, because information about "crossing" of binary builds [from environment](https://github.com/mesonbuild/meson/blob/021d242f9ce8e6cd804af0c1eb4179b8c83fd470/mesonbuild/backend/backends.py#L848). It can be getted from `executable` flag `native` (if `true` test can be runned without wrappers). By now if test `executable` have flag `native: true` it builds for build machine, but [`'Not run because can not execute cross compiled binaries.'`](https://github.com/mesonbuild/meson/blob/c9b7e5bd403971a268530838b25a051ab6310fef/mesonbuild/mtest.py#L642)

**To Reproduce**
`app.cpp`
```cpp
#include <cstdio>
#include "func.h"
int main(int argc, char** argv) { printf("func: %d\n", func(10, 32)); }
```
`func.d`
```d
module func;
extern (C) int func(int a, int b) nothrow @nogc { return a + b; }
unittest { assert (func(10, 12) == 10 + 12); }
```
`func.h`
```cpp
extern "C" int func(int, int);
```
`meson.build`
```
project('app')
add_languages('c', native: false)
add_languages('cpp', native: false)
add_languages('d', native: false)
add_languages('d', native: true)
c_args = ['-fstack-protector-strong', '-fno-plt', '-fno-lto', '-Wno-psabi']
add_global_arguments(c_args, native: false, language: 'c')
add_global_arguments(c_args, native: false, language: 'cpp')
c_link_args = ['-Wl,-O1,-z,relro,-z,now', '-fno-lto']
add_global_link_arguments(c_link_args, native: false, language: 'c')
add_global_link_arguments(c_link_args, native: false, language: 'cpp')
phobos2 = meson.get_compiler('cpp').find_library('phobos2-ldc-shared', static: false, dirs: '/d/ldc2/arm-lib/')
druntime = meson.get_compiler('cpp').find_library('druntime-ldc-shared', static: false, dirs: '/d/ldc2/arm-lib/')
d_dep = declare_dependency(dependencies: [ phobos2, druntime ])
lib = static_library('lib', 'func.d', dependencies: d_dep, native: false)
lib_dep = declare_dependency(link_with: [lib], dependencies: d_dep, include_directories: include_directories('.'))
app_exe = executable('app', 'app.cpp', dependencies: lib_dep, native: false, link_language: 'cpp')
general_test = executable('test_bin', 'func.d', d_unittest: true, native: true, d_args: ['-main'])
test('general', general_test)
```
`cross.ini`
```
[constants]
sysroot = '/sysroot'
toolchain_prefix = '/arm-gcc/bin/arm-linux-gnueabihf-'
dtriple = 'armv7l-linux-gnueabihf'
[properties]
sys_root = sysroot
pkg_config_libdir = sysroot / 'usr/lib/pkgconfig'
[built-in options]
c_args = ['-mcpu=cortex-a7',
          '-mfloat-abi=hard',
          '-mfpu=vfpv4',
          '--sysroot='+sysroot,
          '-L' + sysroot / 'usr/lib'
         ]
cpp_args = c_args
d_args = ['-mtriple='+dtriple ]
cpp_std = 'c++14'
[host_machine]
system = 'linux'
cpu_family = 'arm'
cpu = 'armv7hl'
endian = 'little'
[binaries]
d = 'ldc2'
c = toolchain_prefix + 'gcc'
c_ld = 'gold'
cpp = toolchain_prefix + 'g++'
cpp_ld = 'gold'
ar = toolchain_prefix + 'ar'
strip = toolchain_prefix + 'strip'
pkgconfig = 'pkg-config'
```
steps:

1. `meson _bld --cross-file=cross.ini`
2. `cd _bld && meson test`

result:
```
ninja: Entering directory `/workdir/_bld'
[6/6] Linking target test_bin
1/1 general SKIP           0.00s

Summary of Failures:

1/1 general SKIP           0.00s

Ok:                 0   
Expected Fail:      0   
Fail:               0   
Unexpected Pass:    0   
Skipped:            1   
Timeout:            0   

Full log written to /workdir/_bld/meson-logs/testlog.txt
```
`testlog.txt`
```
...
--- command ---
NONE
--- stdout ---
Not run because can not execute cross compiled binaries.
-------
...
```

**Expected behavior**
Execute native tests on build machine without wrappers.

**system parameters**
* Is this a `cross build`
* fedora 32
* `python --version` is `3.8.6`
* `meson --version` is `0.56.0`
* `ninja --version` is `1.10.1`

### PR DESCRIPTION
**Pull request by deviator**
Need I add tests? How they must be written?

### CODE REVIEW

**Comment by jpakkane** [comment]
You need to add a test to `run_unittests.py` in the cross compilation class, since the test needs to validate that the test is actually run. It should do something like this:

- write a Meson file that builds one executable as `native: true` and runs it as a test
- the test exe should create an empty file, such as `native_test_has_run.stamp` in the current working directory when run
- the test should initialize and build the code
- it should verify that the stamp file does not exist
- it should then run the tests
- it should verify that the stamp file does exists

**Comment by Rogni** [comment]
is there a guideline for writing autotests in meson?

**Comment by Rogni** [comment]
how to run ./run_unittests.py with custom meson.py ?

**Comment by deviator** [comment]
ping @jpakkane

**Comment by jpakkane** [comment]
You should not have merge commits in pull requests. You should remove them and use rebase instead:

https://mesonbuild.com/Contributing.html#keeping-pull-requests-up-to-date

**Comment by deviator** [comment]
ping @jpakkane

File: run_unittests.py [hunk 1]
```diff
@@ -7507,6 +7507,20 @@ def test_pkg_config_option(self):
...
         ])
+    
+    def test_run_native_test(self):
+        '''
+        https://github.com/mesonbuild/meson/issues/7997
+        check run native test in crossbuild without exe wrapper
+        '''
+        testdir = os.path.join(self.unit_test_dir, '88 run native test')
+        stamp_file = os.path.join(self.builddir, 'native_test_has_run.stamp')
+        self.meson_cross_file = os.path.join(testdir, 'cross.txt')
```
Comment (jpakkane) [inline run_unittests.py:7518]: You should not need to have your own cross file. The `setUp` method of `LinuxCrossArmTests` automatically sets this to the common Arm cross compilation file. You should be able to remove this line and your custom cross file as well.

Other than that, looks good.

Comment (deviator) [reply run_unittests.py]: fixed

**Comment by jpakkane** [comment]
Thanks.

**Pull request merged**